source/common/event/file_event_impl.h

```diff
@@ -41,5 +45,119 @@ class FileEventImpl : public FileEvent, ImplBase {
   // polling and activating new fd events.
   const bool activate_fd_events_next_event_loop_;
 };
+
+// Forward declare for friend class.
+class UserSpaceFileEventFactory;
+
+// The interface of populating event watcher and obtaining the active events. The events includes
+// Read, Write and Closed.
+class EventListener {
+public:
+  virtual ~EventListener() = default;
+
+  // Provide the activated events.
+  virtual uint32_t triggeredEvents() PURE;
+  virtual uint32_t getAndClearEphemeralEvents() PURE;
+
+  // Callbacks of the event operation.
+  virtual void onEventEnabled(uint32_t enabled_events) PURE;
+  virtual void onEventActivated(uint32_t activated_events) PURE;
+};
+
+// Return the enabled events except EV_CLOSED. This implementation is generally good since only
+// epoll supports EV_CLOSED but the entire envoy code base supports another poller. The event owner
+// must assume EV_CLOSED is never activated. Also event owner must tolerate that OS could notify
+// events which are not actually triggered.
+class DefaultEventListener : public EventListener {
+public:
+  ~DefaultEventListener() override = default;
+
+  // Return both read and write.
+  uint32_t triggeredEvents() override { return pending_events_ & (~Event::FileReadyType::Closed); }
+
+  void onEventEnabled(uint32_t enabled_events) override { pending_events_ = enabled_events; }
+
+  void onEventActivated(uint32_t activated_events) override {
+    ephermal_events_ |= activated_events;
+  }
+
+  uint32_t getAndClearEphemeralEvents() override {
+    auto res = ephermal_events_;
+    ephermal_events_ = 0;
+    return res;
+  }
+
+private:
+  // The persisted interested events and ready events.
+  uint32_t pending_events_{};
+  // The events set by activate() and will be cleared after the io callback.
+  uint32_t ephermal_events_{};
+};
+
+// A FileEvent implementation which is
+class UserSpaceFileEventImpl : public FileEvent, Logger::Loggable<Logger::Id::io> {
+public:
+  ~UserSpaceFileEventImpl() override {
+    if (schedulable_.enabled()) {
+      schedulable_.cancel();
+    }
+    ASSERT(event_counter_ == 1);
+    --event_counter_;
+  }
+
+  // Event::FileEvent
+  void activate(uint32_t events) override {
+    event_listener_.onEventActivated(events);
```

Comment:
I think (please check) that for libebent, if an event one is not registered for is activated it's a no-op.  I'm not seeing that here.

Reply:
It's probably below? Only support these types of event although libevent defines more.
```
  // Only supported event types are set.
  ASSERT((events & (FileReadyType::Read | FileReadyType::Write | FileReadyType::Closed)) == events);
```

Libevent doesn't require active events be registered. In libevent "registered" is described as "pending"
```
void event_active(struct event *ev, int what, short ncalls);
This function makes an event ev become active with the flags what (a combination of EV_READ, EV_WRITE, and EV_TIMEOUT). The event does not need to have previously been pending, and activating it does not make it pending.
```

Reply:
After PR 11750, FileEventImpl will deliver an activated event that the FD even if it is not part of the mask provided by setEnabled.  We don't tend to activate such events.

The other thing to look at from FileEventImpl, is that setEnable clears ephemeral events.